src/librustc_lint/builtin.rs

```diff
@@ -641,9 +641,23 @@ impl LintPass for UnusedAttributes {
             }
         }
 
+        let plugin_attributes = cx.sess().plugin_attributes.borrow_mut();
+        for &(ref name, ty) in plugin_attributes.iter() {
+            match ty {
+                AttributeType::Whitelisted if attr.check_name(&*name) => {
+                    break;
+                },
+                _ => ()
+            }
+        }
+
         if !attr::is_used(attr) {
             cx.span_lint(UNUSED_ATTRIBUTES, attr.span, "unused attribute");
-            if KNOWN_ATTRIBUTES.contains(&(&attr.name(), AttributeType::CrateLevel)) {
+            if KNOWN_ATTRIBUTES.contains(&(&attr.name(), AttributeType::CrateLevel)) ||
```

Comment:
Now that this is multiple lines, could you break (parts of) it out into separate variables?